nemo/collections/nlp/models/duplex_text_normalization/duplex_tagger.py

```diff
@@ -0,0 +1,315 @@
+        passed in as `batch`.
+        """
+        self.train()
+        num_labels = self.num_labels
+
+        # Apply Transformer
+        input_ids = batch['input_ids'].to(self.device)
+        input_masks = batch['attention_mask'].to(self.device)
+        tag_logits = self.model(input_ids, input_masks).logits
+
+        # Loss
+        tag_labels = batch['labels'].view(-1).to(self.device)
+        train_loss = self.loss_fct(tag_logits.view(-1, num_labels), tag_labels)
+
+        lr = self._optimizer.param_groups[0]['lr']
+        self.log('train_loss', train_loss)
+        self.log('lr', lr, prog_bar=True)
+        return {
+            'loss': train_loss,
+            'lr': lr
+        }
+
+    # Validation and Testing
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        self.eval()
+        num_labels = self.num_labels
+
+        # Apply Transformer
+        input_ids = batch['input_ids'].to(self.device)
+        input_masks = batch['attention_mask'].to(self.device)
+        tag_logits = self.model(input_ids, input_masks).logits
+        tag_preds = torch.argmax(tag_logits, dim=2)
+
+        # Loss
+        tag_labels = batch['labels'].to(self.device)
+        val_loss = self.loss_fct(tag_logits.view(-1, num_labels),
+                                 tag_labels.view(-1))
+
+        # Extract batch_predictions and batch_labels
+        predictions, labels = tag_preds.tolist(), tag_labels.tolist()
+        final_predictions = [
+            [constants.ALL_TAG_LABELS[p] for (p, l) in zip(prediction, label) \
+             if l != constants.LABEL_PAD_TOKEN_ID]
+            for prediction, label in zip(predictions, labels)
+        ]
+        final_labels = [
+            [constants.ALL_TAG_LABELS[l] for (p, l) in zip(prediction, label) \
+             if l != constants.LABEL_PAD_TOKEN_ID]
+            for prediction, label in zip(predictions, labels)
+        ]
+        self.val_all_preds.extend(final_predictions)
+        self.val_all_targets.extend(final_labels)
+
+        return {
+            'val_loss': val_loss
+        }
+
+    def validation_epoch_end(self, outputs):
+        """
+        Called at the end of validation to aggregate outputs.
+        :param outputs: list of individual outputs of each validation step.
+        """
+        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
+
+        self.log('val_loss', avg_loss)
+
+        # Compute sentence_accuracy
+        sent_count, sent_correct = 0, 0
+        for ix, (p, l) in enumerate(zip(self.val_all_preds, self.val_all_targets)):
+            # Update stats
+            sent_correct += int(p==l)
+            sent_count += 1
+        sent_accuracy = sent_correct / sent_count
+        self.log('val_sentence_accuracy', sent_accuracy)
```

Comment:
better use pytorch lightning metrics to do this, otherwise it will probably not be allreduced

Reply:
Yes I fixed that.